### PR TITLE
[JENKINS-34395] Support for tag discovery

### PR DESCRIPTION
**Pull request by stephenc**
See [JENKINS-34395](https://issues.jenkins-ci.org/browse/JENKINS-34395)

I believe this to be a full implementation, but test coverage of some aspects of this is non-trivial... OTOH given that it is mostly new functionality that would be needing testing we could just go as is

Downstream of https://github.com/kohsuke/github-api/pull/375

@reviewbybees

### CODE REVIEW

**Comment by stephenc** [comment]
Screenshots:

Configuring for tag discovery:

screen shot 2017-09-01 at 14 41 01

Indexing log when tags are being discovered:
screen shot 2017-09-01 at 14 41 29

Results of tag discovery:
screen shot 2017-09-01 at 14 41 43

**Comment by HRMPW** [comment]
How can you limit what Tags are supported, @stephenc ? If I have a repo with 500 tags I don't want 500 new projects created.

Should this be limited to annotated tags? Can it skip "Pre-release" tags?

**Comment by stephenc** [comment]
An extension plugin can trivially filter tags by date created

**Comment by stephenc** [comment]
There are several example plugins doing filtering and TagSCMHead.getDate() is the last commit for shallow tags and the tag date for annotated tags

(We could add an option to discover annotated tags only/lightweight only/signed only/etc... or maybe make that a filtering concern)

**Comment by HRMPW** [comment]
@stephenc We would want an Environment variable or something that can be used in a `when` block to control how a Tag run could be customized in the Pipeline.

@orrc creates two Environment Variables in his Git Tag Message Plugin. Something like that would help. https://wiki.jenkins.io/display/JENKINS/Git+Tag+Message+Plugin

@michaelneale @i386 @vivek @abayer Many users have been requesting this functionality. Currently, it wouldn't show up in Blue Ocean. 

We have had several discussions about how this relates to creating a release or how it might work with a mvn tag. If someone creates a Pipeline with "mvn release" it will create a new tag. This will either create a new job to run if Tags are enabled or it will not show up in the Tags tab like jobs created by Tags.

**Comment by stephenc** [comment]
@HRMPW 
> We would want an Environment variable or something that can be used in a when block to control how a Tag run could be customized in the Pipeline.

you mean something like https://github.com/jenkinsci/branch-api-plugin/blob/master/src/main/java/jenkins/branch/BranchNameContributor.java#L88-L91 (i.e. with this change you get `TAG_NAME`, `TAG_TIMESTAMP`, `TAG_DATE` and `TAG_UNIXTIME` right now as long as you are using branch-api 2.0.11 or newer

**Comment by stephenc** [comment]
> Currently, it wouldn't show up in Blue Ocean.

Of course if BlueOcean had actually implemented support for SCMHeadCategories, then it would magically just show up... but [JENKINS-40367](https://issues.jenkins-ci.org/browse/JENKINS-40367) hasn't had much traction

**Comment by stephenc** [comment]
> Currently, it wouldn't show up in Blue Ocean.

I suspect given that BlueOcean is ignorant of categories, tags will show up under branches.

Lack of use of categories is also likely the root cause of slow rendering of pull requests tab in BlueOcean that I've seen JIRA for (but on phone I cannot find link for)

**Comment by stephenc** [comment]
There is also a similar PR for the git-plugin

**Comment by michaelneale** [comment]
@stephenc yeah there is some old code there. @vivek  has a todo on the PR thing so he may take a look at categories etc as part of that as refresh at some point.

**Comment by vivek** [comment]
@stephenc 

> I suspect given that BlueOcean is ignorant of categories, tags will show up under branches.

blueocean (for GitHub/bitbucket) uses [BranchDiscoveryTrait](https://github.com/jenkinsci/blueocean-plugin/blob/76af0a305b454fba4bff4bb687d2b26614b87496/blueocean-github-pipeline/src/main/java/io/jenkins/blueocean/blueocean_github_pipeline/GithubPipelineCreateRequest.java#L51) with strategyId 3 to take in all branches - that is all PRs and branches. Looking at this trait, tags won't show up in blueocean without extra work that is do it properly in JENKINS-40367.

**Comment by stephenc** [comment]
@vivek what I mean is that if the user configured the multibranch project to have the tag discovery trait then they will likely show up as branches in the blue ocean UI

**Comment by witokondoria** [comment]
@HRMPW @stephenc im planning on submitting a [date filtering trait](https://github.com/witokondoria/git-aged-refs-trait-plugin) once bitbucket-branch-source-plugin#67 gets merged and released

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 1]
```diff
@@ -987,6 +1046,216 @@ public SCMRevision create(@NonNull PullRequestSCMHead head,
...
+            } catch (WrappedException e) {
+                try {
+                    e.unwrap();
+                } catch (RateLimitExceededException rle) {
+                    throw new AbortException(rle.getMessage());
+                }
+            }
+            return result;
+        } finally {
+            Connector.release(github);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: Can this not become `[Auto]Closeable` _[sic]_?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java [hunk 2]
```diff
@@ -987,6 +1046,216 @@ public SCMRevision create(@NonNull PullRequestSCMHead head,
...
+            if (StringUtils.isBlank(repository)) {
+                throw new AbortException("No repository selected, skipping");
+            }
+
+            String fullName = repoOwner + "/" + repository;
+            ghRepository = github.getRepository(fullName);
+            final GHRepository ghRepository = this.ghRepository;
+            GitHubSCMSourceContext context = new GitHubSCMSourceContext(null, SCMHeadObserver.none())
+                    .withTraits(traits);
+            Matcher prMatcher = Pattern.compile("^PR-(\\d+)(-.*)?$").matcher(headName);
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/GitHubSCMSource.java]: Is there not already some kind of constant for this?

File: src/main/java/org/jenkinsci/plugins/github_branch_source/PushGHEventSubscriber.java [hunk 3]
```diff
@@ -308,6 +322,39 @@ public String getSourceName() {
...
+                // then the event consumer is supposed to turn arround and do a fetch(..., event, ...)
+                // and as GitHubSCMSourceRequest strips out the timestamp in calculating the requested
+                // tag names, we have a winner.
+                //
+                // So let's make the assumption that tags are not pushed a long time after their creation and
+                // use the event timestamp. This may cause issues if anyone has a pre-filter that filters
+                // out tags that are less than X seconds old, but as such a filter would be incompatible with events
+                // discovering tags, no harm... the key part is that a pre-filter that removes tags older than X days
+                // will not strip the tag *here* (because it will always be only a few seconds "old"), but when
+                // the fetch call actually has the real tag date the pre-filter will apply at that point in time.
```
Comment (jglick) [inline src/main/java/org/jenkinsci/plugins/github_branch_source/PushGHEventSubscriber.java]: Uh, OK. I guess this is all hypothetical unless and until someone writes such a filter.

**Comment by stephank** [comment]
I'm not well versed in Java-land, but have deployed this to our Jenkins install, and found two things so far:

 - The `org.kohsuke.github-api` dep of the github-api-plugin needs to be updated, because the current 1.86 is missing the `listRefs(String)` overload
 - I believe this fails on repo's with no tags at all:

```
Sep 19, 2017 11:19:51 AM jenkins.branch.OrganizationFolder$OrganizationScan run
INFO: <redacted> #20170919.111643 organization scan action completed: NOT_BUILT in 3 min 8 sec
Sep 19, 2017 11:19:51 AM hudson.model.Executor finish1
SEVERE: Executor threw an exception
java.lang.Error: org.kohsuke.github.GHFileNotFoundException: {"message":"Not Found","documentation_url":"https://developer.github.com/v3"}
        at org.kohsuke.github.Requester$PagingIterator.fetch(Requester.java:516)
        at org.kohsuke.github.Requester$PagingIterator.hasNext(Requester.java:481)
        at org.kohsuke.github.PagedIterator.fetch(PagedIterator.java:44)
        at org.kohsuke.github.PagedIterator.hasNext(PagedIterator.java:32)
        at org.jenkinsci.plugins.github_branch_source.GitHubSCMSource.retrieve(GitHubSCMSource.java:982)
        at jenkins.scm.api.SCMSource._retrieve(SCMSource.java:355)
        at jenkins.scm.api.SCMSource.fetch(SCMSource.java:309)
        at jenkins.branch.MultiBranchProjectFactory$BySCMSourceCriteria.recognizes(MultiBranchProjectFactory.java:263)
        at jenkins.branch.OrganizationFolder$SCMSourceObserverImpl$1.recognizes(OrganizationFolder.java:1297)
        at jenkins.branch.OrganizationFolder$SCMSourceObserverImpl$1.complete(OrganizationFolder.java:1312)
        at jenkins.scm.api.trait.SCMNavigatorRequest.process(SCMNavigatorRequest.java:256)
        at jenkins.scm.api.trait.SCMNavigatorRequest.process(SCMNavigatorRequest.java:206)
        at org.jenkinsci.plugins.github_branch_source.GitHubSCMNavigator.visitSources(GitHubSCMNavigator.java:961)
        at jenkins.branch.OrganizationFolder.computeChildren(OrganizationFolder.java:412)
        at com.cloudbees.hudson.plugins.folder.computed.ComputedFolder.updateChildren(ComputedFolder.java:276)
        at com.cloudbees.hudson.plugins.folder.computed.FolderComputation.run(FolderComputation.java:165)
        at jenkins.branch.OrganizationFolder$OrganizationScan.run(OrganizationFolder.java:863)
        at hudson.model.ResourceController.execute(ResourceController.java:97)
        at hudson.model.Executor.run(Executor.java:419)
Caused by: org.kohsuke.github.GHFileNotFoundException: {"message":"Not Found","documentation_url":"https://developer.github.com/v3"}
        at org.kohsuke.github.Requester.handleApiError(Requester.java:686)
        at org.kohsuke.github.Requester$PagingIterator.fetch(Requester.java:510)
        ... 18 more
Caused by: java.io.FileNotFoundException: <redacted>
        at com.squareup.okhttp.internal.huc.HttpURLConnectionImpl.getInputStream(HttpURLConnectionImpl.java:243)
        at com.squareup.okhttp.internal.huc.DelegatingHttpsURLConnection.getInputStream(DelegatingHttpsURLConnection.java:210)
        at com.squareup.okhttp.internal.huc.HttpsURLConnectionImpl.getInputStream(HttpsURLConnectionImpl.java:25)
        at org.kohsuke.github.Requester.parse(Requester.java:612)
        at org.kohsuke.github.Requester.parse(Requester.java:594)
        at org.kohsuke.github.Requester.access$200(Requester.java:73)
        at org.kohsuke.github.Requester$PagingIterator.fetch(Requester.java:505)
        ... 18 more
```

**Comment by stephenc** [comment]
@stephank Can you give the update in 4d47488 a try and comment back, should fix those two issues

**Comment by stephank** [comment]
@stephenc The issues are fixed, and scanning now works!

I'm still missing something in configuration, though. Scan log now reads:

```
    Checking tag 0.0.1
      ‘Jenkinsfile’ found
    Met criteria
No automatic builds for 0.0.1
```

Which leads me to believe I'm not passing this check: https://github.com/jenkinsci/branch-api-plugin/blob/c0a695746af764b5a917891860dd4ae391fbdbc5/src/main/java/jenkins/branch/MultiBranchProject.java#L2144-L2174

So tags are skipped by default. But I'm not sure where such a strategy is configured, so I can override that behavior?

For what it's worth, this is on a github organisation item. I am running current current master of this plugin and the github-api plugin only. Not sure if I need to run a more recent branch-api plugin too?

Thanks for all the work you're putting into this, btw. 👍

**Comment by stephenc** [comment]
> No automatic builds for 0.0.1

that is functioning exactly as designed. You would need an extension plugin that defines a `BranchBuildStrategy` that allows building tags... this could be as simple as the test of that code uses: https://github.com/jenkinsci/branch-api-plugin/blob/28be8058a6faea90f0549b80cce2c24714f69ea4/src/test/java/integration/EventsTest.java#L589-L600 in other words:

```
 public class BuildEverythingStrategyImpl extends BranchBuildStrategy {
        @Override
        public boolean isAutomaticBuild(SCMSource source, SCMHead head) {
            return true;
        }

        @Extension
        public static class DescriptorImpl extends BranchBuildStrategyDescriptor {
            public String getDisplayName() {
                 return "Build everything";
            }
        }
    }
```

But you probably want something that automatically builds change requests, regular branches and for tags only automatically builds if it is "recent" (for some definition of recent - probably user configurable).

**Comment by stephenc** [comment]
You should be able to manually trigger a build of a tag without a `BranchBuildStrategy` though

**Comment by stephank** [comment]
Check! Did a manual build, checks out the correct tree and works nicely.

Awesome, and thanks again! :D

**Comment by stephenc** [comment]
Also in terms of composability, it might be better to have at least three `BranchBuildStrategy` implementations... one which does: 

```
public boolean isAutomaticBuild(SCMSource source, SCMHead head) {
    return head instanceof ChangeRequestSCMHead;
}
```

one which does

```
public boolean isAutomaticBuild(SCMSource source, SCMHead head) {
    // auto-build for regular branches
    return UncategorizedSCMHeadCategory.DEFAULT.isMatch(head,source.getCategories());
}
```

and a final one:

```
public boolean isAutomaticBuild(SCMSource source, SCMHead head) {
    // this will auto-build tags created in the last week only
    return head instanceof TagSCMHead && ((TagSCMHead)head).getTimestamp() > System.currentTimeMillis()-TimeUnit.DAYS.toMillis(7);
}
```

**Comment by stephenc** [comment]
@stephank for the record, I am trying to keep the core plugins with only generic "sensible" behaviours, hence why branch-api will not build tags by default (because people may have configured their `Jenkinsfile` to deploy to production if the head is a Tag, and if you recreate a new multibranch and it goes and discovers all your old tags... production would be toast)

What I want to encourage and enable is people who have specific needs to write their own extension plugins for the specific behaviour they need... so if you need tags to build automatically, I want you to be able to write a branch-api extension plugin that meets your needs... and anyone else who has that need can then install your plugin and get the functionality... but users who do not need it are not distracted with UI configuration options and requirements that are really an aside from their generic needs

**Comment by stephank** [comment]
@stephenc That makes sense. The config pages are already complicated enough. :)

I'm not sure if there's more you want to do to avoid confusion? Right now, 'Discover Tags' creates the items, but doesn't build. Perhaps a note that explains the default build strategy skips tags, with eventually a link to a suggested plugin? (But the same confusion happens in fetch behaviours.)

I tried building a small plugin with just the extension doing `return true`, but also cannot seem to find anywhere where the build strategy is selected. There's nothing on the github org config page, apparently?

**Comment by mtb-xt** [comment]
An example of how to build such a plugin would be really nice >_<

**Comment by stephenc** [comment]
@mtb-xt 

You want a plugin that depends on branch-api (and probably scm-api and - with test scope - the scm-api test harness)

Then you will start with just one class:

```
public class BuildEverythingStrategyImpl extends BranchBuildStrategy {
        @Override
        public boolean isAutomaticBuild(SCMSource source, SCMHead head) {
            return true;
        }

        @Extension
        public static class DescriptorImpl extends BranchBuildStrategyDescriptor {
            public String getDisplayName() {
                 return "Build everything";
            }
        }
    }
```

And a matching `config.jelly` 

```
<?jelly escape-by-default='true'?>
<j:jelly xmlns:j="jelly:core" />
```

And I'd add a test that is basically a copy of https://github.com/jenkinsci/branch-api-plugin/blob/28be8058a6faea90f0549b80cce2c24714f69ea4/src/test/java/integration/EventsTest.java#L602-L635

I wouldn't release that plugin, but rather use that as the starting point to develop the build strategies you want.

So you will have

```
pom.xml
src/main/java/.../BuildEverythingStrategyImpl.java
src/main/resources/index.jelly
src/main/resources/.../BuildEverythingStrategyImpl/config.jelly
src/test/java/.../BuildEverythingStrategyImplTest.java
```

The `index.jelly` will be

```
<?jelly escape-by-default='true'?>
<div>
A plugin that provides a branch build strategy that builds everything automatically for use in job types that are based on Branch API such as Multibranch Projects and Organization Folders
</div>
```

And I'll make it even easier, here's the `pom.xml`

```
<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
  <modelVersion>4.0.0</modelVersion>

  <parent>
    <groupId>org.jenkins-ci.plugins</groupId>
    <artifactId>plugin</artifactId>
    <version>2.36</version>
    <relativePath />
  </parent>

  <artifactId>branch-build-strategy-everything-impl</artifactId>
  <version>0.1-SNAPSHOT</version>
  <packaging>hpi</packaging>

  <name><!-- your plugin name --></name>
  <description>
    <!-- the same text as between the divs in your index.jelly, no HTML tags though -->
  </description>
  <url><!-- URL of the wiki page you create for your plugin --></url>
  <licenses>
    <!-- I recommend MIT for Jenkins plugins, but pick whatever you want -->
    <license>
      <name>The MIT license</name>
      <url>https://opensource.org/licenses/MIT</url>
      <distribution>repo</distribution>
    </license>
  </licenses>

  <developers>
    <developer>
      <id>your-jenkins-jira-username</id>
      <name>Your Name</name>
    </developer>
  </developers>

  <scm>
    <connection>scm:git:git://github.com/jenkinsci/${project.artifactId}-plugin.git</connection>
    <developerConnection>scm:git:git@github.com:jenkinsci/${project.artifactId}-plugin.git</developerConnection>
    <url>http://github.com/jenkinsci/${project.artifactId}-plugin</url>
    <tag>HEAD</tag>
  </scm>

  <properties>
    <jenkins.version><!-- pick a version, needs to be at least as high as your dependencies,
 you can go with 1.642.3, but I would recommend at least the 2.32's for a new plugin--></jenkins.version>
    <scm-api.version>2.2.2</scm-api.version>
  </properties>

  <repositories>
    <repository>
      <id>repo.jenkins-ci.org</id>
      <url>https://repo.jenkins-ci.org/public/</url>
    </repository>
  </repositories>

  <pluginRepositories>
    <pluginRepository>
      <id>repo.jenkins-ci.org</id>
      <url>https://repo.jenkins-ci.org/public/</url>
    </pluginRepository>
  </pluginRepositories>

  <dependencies>
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>scm-api</artifactId>
      <version>${scm-api.version}</version>
    </dependency>
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>branch-api</artifactId>
      <version>2.0.11</version>
    </dependency>
    <!-- test dependencies -->
    <dependency>
      <groupId>org.jenkins-ci.plugins</groupId>
      <artifactId>scm-api</artifactId>
      <version>${scm-api.version}</version>
      <classifier>tests</classifier>
      <scope>test</scope>
    </dependency>
  </dependencies>
</project>
```

This is really that easy... give it a go... write your first Jenkins plugin

**Comment by stephenc** [comment]
@stephank Hmmm is it available for a regular multibranch project?

If available for a regular multibranch project but not for an org folder then that is a bug, please file a JIRA and we can see about addressing that relatively easily... I'd even be happy to help you create the PR for that (against branch-api) if you are interested.

If your extension is not configurable in multibranch projects, then likely your extension is not implemented correctly. If you can share your code I'd be happy to review to help identify where the problem might be (assuming your extension is not configurable in regular multibranch projects... which it should be, see https://github.com/jenkinsci/branch-api-plugin/blob/master/src/main/resources/jenkins/branch/BranchSource/config.jelly#L38-L49)

**Comment by stephank** [comment]
I just came up with the idea to always `return false`, and see if that'd reject automated builds in a release version of github-branch-source-plugin and regular old `Discover Branches`. But it doesn't, and it appears the entry in "View Configuration" of a repo is just something broken.

**Comment by mtb-xt** [comment]
I can report the same thing, the configuration window for a created repository under GutHub folders - can't load, it hangs on 'loading'  screen. Looks like it lacks some message files to populate the options. 
I can't find that option under github organisation folder settings menu too.

**Comment by stephenc** [comment]
@mtb-xt can you find a stack trace in the logs co-incident with opening the page?

Can you find a JavaScript console log error loading the config page?

**Comment by mtb-xt** [comment]
I can only see this in the logs when I open the page:

```
Oct 02, 2017 8:58:06 AM hudson.ExtensionFinder$GuiceFinder$FaultTolerantScope$1 error
INFO: Failed to instantiate optional component com.github.mjdetullio.jenkins.plugins.multibranch.IvyBranchProjectFactory$DescriptorImpl; skipping
```

I can't figure out which logger to use to capture more detailed log. My plugin's class name does not appear there. Could you direct me somewhere on how to log for my plugin properly?

This is the javascript console:  https://gist.github.com/mtb-xt/7345d32ae1f7614f89b903c2d30e5219

**Comment by stephenc** [comment]
@stephank / @mtb-xt w.r.t. the issue on BranchBuildStrategy I have filed JENKINS-47311 and the fix https://github.com/jenkinsci/branch-api-plugin/pull/113 if you care to review

**Comment by mtb-xt** [comment]
@stephenc thanks for the fix, it does fix the config issue on Organisation Folder page - at least partially - I can see a selector to add build strategies. But, when I click save or apply, I'm getting a null pointer exception - at jenkins.branch.OrganizationFolder.submit(OrganizationFolder.java:320) - full strack trace here - https://gist.github.com/mtb-xt/8c12a3090053d6f36acc97f7161cda56. Note, that this happens both when I have my build strategy enabled or disabled, and even if I delete 'discover tags' option  at all.

**Comment by stephenc** [comment]
@mtb-xt https://issues.jenkins-ci.org/browse/JENKINS-47340 filed and 2.0.13 of branch-api released. You can download it from https://jenkins-updates.cloudbees.com/download/plugins/branch-api/2.0.13/branch-api.hpi if you want to test ahead of being picked up by the oss update centre

**Comment by mtb-xt** [comment]
@stephenc I've updated, it does fix the null pointer thing. I've found another issue - Organisation folder configuration page won't remember added build strategy - it adds ok, if I press 'apply' it says 'saved'. If I revisit the configuration page - the build strategy isn't there. Same thing happens if I press 'save'.
If I do the same thing from configuration page for a repository inside Organisation folder, and add a build strategy - it won't even show me the 'save' and 'apply' buttons <_<

**Comment by stephenc** [comment]
@mtb-xt 13 is an unlucky version number. 2.0.14 released

**Comment by stephenc** [comment]
> If I do the same thing from configuration page for a repository inside Organisation folder, and add a build strategy - it won't even show me the 'save' and 'apply' buttons <_<

Functions as designed. That is the "View Configuration" screen as the children of an Organization folder are managed by the organization folder, you do not have the ability to configure them independently

**Comment by WoLfulus** [comment]
@stephenc I'm evaluating Jenkins for a project, and this is the the only issue I found so far. To install and test this I should just copy the .hpi into plugins folder and restart Jenkins?

**Comment by stephenc** [comment]
@WoLfulus just waiting on @i386 to confirm ok from BlueOcean for merge and release

**Comment by WoLfulus** [comment]
Awesome! Thank you!

**Comment by mtb-xt** [comment]
@stephenc I've successfully tested a 'build everything' strategy plugin, seems to be working! Thanks for the help.

I still have some errors relating to tags in my jenkins log
hudson.plugins.git.GitException: Error launching git whatchanged
and
com.google.common.cache.CacheLoader$InvalidCacheLoadException: CacheLoader returned null for key production-2017101.

Full stack traces here - https://gist.github.com/mtb-xt/160054019efbcf22166bbeb527ff33c2

**Comment by WoLfulus** [comment]
@stephenc I've compiled the plugin and installed it for testing. Everything is good, but I faced a problem that I'm not really sure if its caused by this plugin or its dependents.

When I tried pushing a branch with slash "feature/random-branch" Jenkins broke.

```
[-test_feature_random-branch-IGZEUUURSCIYMU43FFKLTP2EIE5NEEI2QTBUUT6UBGH77NFZKMOQ] Running shell script
+ docker build -t idwall/random-test .
unable to prepare context: unable to evaluate symlinks in Dockerfile path: lstat /var/lib/jenkins/workspace/-test_feature_random-branch-IGZEUUURSCIYMU43FFKLTP2EIE5NEEI2QTBUUT6UBGH77NFZKMOQ/Dockerfile: no such file or directory
```

The project name is `random-test` and the branch name is `feature/random-branch`

Comparing with other branches, it seems that the folder name is wrong (missing part of the project name for some reason.

I'm asking if it could be something related to this plugin because we do have other pipelines that uses gitflow (branches with slashes) and we haven't had problems with it yet (they don't use github branch source).

**Comment by stephenc** [comment]
@WoLfulus the slash stuff is unrelated, I suggest you bring that up on the jenkins-users mailing list.

w.r.t. the FileNotFoundException, did you check the github api url? `https://api.github.com/repos/organization/random-test/git/refs/tags/v1.0.0` I checked for example: `https://api.github.com/repos/jenkinsci/github-branch-source-plugin/git/refs/tags/github-branch-source-2.2.3` and that exists, so I cannot see how you would get a 404 from that url if - as you say - the tag has been pushed to github, unless the credentials you are using do not have permssion somehow.

**Comment by stephenc** [comment]
@WoLfulus hmm are your tags lightweight tags or annotated tags? `github-branch-source-2.2.3` is an annotated tag... 

Nope... just verified with a temporary lightweight tag:

lightweight tag

contrast with an annotated tag, where the object type is `tag` and not `commit`

annotated tag

but neither gave a 404 (the temp-v1 tag has now been deleted, so now it will give a 404, but when it was there it did not)

**Comment by WoLfulus** [comment]
@stephenc the slash thing I think its because theres a character limit and it changes from tag to tag depending on the length

About the tag url: requesting by hand gives me 404 too. maybe authentication issues?

**Comment by chris5287** [comment]
@stephenc Any updates on when this might be merged/release?

**Comment by stephenc** [comment]
@chris5287 we are awaiting @i386 / @michaelneale to sign-off

**Comment by michaelneale** [comment]
@stephenc hoping for a blessing from Karl (not sure of his github nick) or the blue smoke to go up soon.

**Comment by chris5287** [comment]
@stephenc thanks for the update. How easy would it be to get these changes ported over to the Bitbucket plugin as well?

**Comment by stephenc** [comment]
@chris5287 if you want to give it a try i’d be happy to review/mentor. I’m busy on a different project so I don’t have time to implement myself

**Comment by kshultzCB** [comment]
@michaelneale This is my GitHub nick. :)
I've got an ATH run going on this as we speak.
--Karl

**Comment by cleverkidd** [comment]
@stephenc
Hi Stephen,
I created the plugin and deployed it to jenkins. The new option appeared as expected `Build everything`. If I add the option and click `save` button, i get following exception: https://pastebin.com/nvbnShiX

can you please advise?
my jenkins-ci.plugins version is 2.36
jenkins version 2.46.3
scm-api version 2.2.3
branch-api version 2.0.14

the code is the same as you provided here

thanks in advance

**Comment by ahammond** [comment]
+1 the ability to trigger builds when a new tag arrives would be a huge win for us.

**Comment by stephenc** [comment]
@cleverkidd I created [JENKINS-47678](https://issues.jenkins-ci.org/browse/JENKINS-47678) to track that stack trace you identified. That issue is fixed in branch-api 2.0.15 (which should be available from the update centers in the next few hours)

**Comment by jeff-knurek** [comment]
I feel like I'm missing something on how this is suppose to work.
I've got 
branch-api 2.0.15 , 
github-branch-source-plugin 2.2.5,
and https://github.com/AngryBytes/jenkins-build-everything-strategy-plugin installed

In the org folder setup, I've added the `Build everything` strategy. I push a tag to the repo that's being watched and nothing happens. I run Scan repository now, and the logs mention nothing about tags:
```Examining org/repo

  Checking branches...

  Getting remote branches...

    Checking branch feature/publishers
      ‘Jenkinsfile’ found
    Met criteria
No changes detected: feature/publishers (still at a29373c2e692bf549f7ff114e4be6ed82f7d056d)

    Checking branch master
      ‘Jenkinsfile’ found
    Met criteria
No changes detected: master (still at a9bc55ee5d6e82111f6ca70ea6168da9289ddd12)

    Checking branch test
      ‘Jenkinsfile’ found
    Met criteria
No changes detected: test (still at b224b0276df138020efed0d1545f97c4bff294cd)

  3 branches were processed

  Checking pull-requests...

  Getting remote pull requests...

  0 pull requests were processed

Finished examining org/repo```

**Comment by stephenc** [comment]
@jeff-knurek

> I feel like I'm missing something on how this is suppose to work.
I've got
branch-api 2.0.15 ,
github-branch-source-plugin 2.2.5,

So you would need to have 2.2.6-SNAPSHOT where you build this plugin with this PR merged.

(Hint: this plugin has not been merged yet)

**Comment by jeff-knurek** [comment]
I pulled the latest code from this repo, then rebased the PR code with master, built the hpi, `2.2.6-SNAPSHOT (private-ed2a6390-root)`
where `ed2a6390` is a reference to the last commit of this PR: https://github.com/jenkinsci/github-branch-source-plugin/pull/158/commits/ed2a6390

But still not seeing the mention of tags during a Scan??

**Comment by stephenc** [comment]
@jeff-knurek did you add “Discover Tags” behaviour?

**Comment by jeff-knurek** [comment]
> @jeff-knurek did you add “Discover Tags” behaviour?
that's it!! thanks

I looked for that on my first build of the plugin, but it was still missing the change, and didn't look for it again after

**Comment by kshultzCB** [comment]
I've run any number of manual and automated tests against this PR, the changes look good to me. 

I did observe some typical ATH flakiness, and one of the resulting issues is [JENKINS-47702](https://issues.jenkins-ci.org/browse/JENKINS-47702). This could be recreated without this PR in place.

**Comment by ahammond** [comment]
Yay!

**Comment by ahammond** [comment]
How would I get this on my jenkins server? We're running a container FROM the standard docker container that pulls plugins at build. I'm guessing I just rebuild our container?

**Comment by stephenc** [comment]
@ahammond I have used my psychic powers to determine exactly how your instance is configured and maintained. While I was doing that I took the liberty of upgrading the plugin for you... eh damn... seems my psychic powers are not working today

**Comment by ahammond** [comment]
@stephenc The correct answer turns out to be "yeah, if you're pulling plugins at build, you should get the latest version of this plugin the next time you build". Tags are now building for us. Thanks.

**Comment by ashb** [comment]
A question about the "build everything plugin" idea. Does the tag fetching distinguish between the initial scan (which might pull in a lot of tags) and subsequent polls/updates/events?

i.e. is it possible for me to write a plugin that will build only newly created tags, but not any tags found during the intial scan?

(i.e. like to have my Jenkins instance be re-createable, and i don't persist any of the workspace or files from instance to instance.)

As a backup I could use the date of the commit the tag is on (we use lightweight tags usually) and just ignore any tag not created in the last X*2 mins (where X is the check interval).

**Comment by stephenc** [comment]
@ashb this is entirely up to the person writing the BranchBuildStrategy extension... you can have multiple such ones. Build Everything is just the most basic one.

I recommend that there would be three or more strategies for building branches, tags and PRs. The tag one could be either simple or it could allow configuration to build based on the tag date (the TagSCMHead mixin has a tag date)

**Comment by ashb** [comment]
Sorry, I meant does a strategy plugin have enough info to distinguish those two cases? (first time populating repo vs new tag in a repo that has been scanned before)

**Comment by stephenc** [comment]
how could it? all it knows is there is now a tag where once there was not and the tag reports a tag date of XYZ (depending on the type of tag that could be the last commit time or the time the heavyweight tag was created)

**Comment by stephenc** [comment]
but you could have the strategy say "only build tags that are less than X days old and must be after today the time I am saving the form"

**Comment by stephenc** [comment]
This issue is merged. There is are mailing lists for discussion

**Pull request merged**